src/node_url.cc

```diff
@@ -1351,6 +1351,53 @@ namespace url {
                             v8::NewStringType::kNormal).ToLocalChecked());
   }
 
+  static void ToUSVString(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+    CHECK_GE(args.Length(), 1);
+    Local<String> str;
+    if (!args[0]->ToString(env->context()).ToLocal(&str))
+      return;
+
+    // The algorithm below can be replaced with `Utf8Value`, which
+    // automatically replaces invalid Unicode surrogates with U+FFFD
+    // REPLACEMENT CHARACTER as the Web IDL spec mandates through V8's
+    // `String::REPLACE_INVALID_UTF8`. However, using `Utf8Value` would prevent
+    // us from using the `changed` optimization, and V8 would have to create a
+    // new string even when it is not needed.
+
+    TwoByteValue value(env->isolate(), str);
+    const size_t n = value.length();
+    bool changed = false;
+
+    for (size_t i = 0; i < n; i++) {
+      uint16_t c = value[i];
+      if (c < 0xD800 || c > 0xDFFF) {
+        continue;
+      } else if (0xDC00 <= c || i == n - 1) {
+        changed = true;
+        value[i] = 0xFFFD;
+      } else {
+        uint16_t d = value[i + 1];
+        if (0xDC00 <= d && d <= 0xDFFF) {
+          i++;
+        } else {
+          changed = true;
+          value[i] = 0xFFFD;
+        }
+      }
+    }
```

Comment:
> I'll benchmark this, but is a regex faster than character search these days?

I expect the answer is 'yes' for short strings.  Calling into C++ land and copying out the string has quite a bit of overhead associated with it.  It's amortized for long strings but it's probably quite substantial for short strings (and I assume most inputs will be relatively short.)

> Also, the regex is more like /([^\uD800-\uDBFF]|^)[\uDC00-\uDFFF]|[\uD800-\uDBFF]([^\uDC00-\uDFFF]|$)/, though if there's a way to simplify that will probably help performance-wise as well.

A simple scan for `/[\uD800-\uDFFF]/` followed by a call into C++ land when it tests true may already be good enough.